lib/src/index.rs

```diff
@@ -28,49 +29,73 @@ use crate::operation::Operation;
 use crate::revset::{ResolvedExpression, Revset, RevsetEvaluationError};
 use crate::store::Store;
 
-/// Error while reading index from the `IndexStore`.
+/// Returned if an error occurs while reading an index from the [`IndexStore`].
 #[derive(Debug, Error)]
 #[error(transparent)]
 pub struct IndexReadError(pub Box<dyn std::error::Error + Send + Sync>);
 
-/// Error while writing index to the `IndexStore`.
+/// Returned if an error occurs while writing an index to the [`IndexStore`].
 #[derive(Debug, Error)]
 #[error(transparent)]
 pub struct IndexWriteError(pub Box<dyn std::error::Error + Send + Sync>);
 
-/// Error to be returned if `Index::all_heads_for_gc()` is not supported by the
+/// An error returned if `Index::all_heads_for_gc()` is not supported by the
 /// index backend.
 #[derive(Debug, Error)]
 #[error("Cannot collect all heads by index of this type")]
 pub struct AllHeadsForGcUnsupported;
 
+/// Defines the interface for types that provide persistent storage for an
+/// index.
 pub trait IndexStore: Send + Sync + Debug {
+    #[allow(missing_docs)]
     fn as_any(&self) -> &dyn Any;
 
+    /// Returns a name representing the type of index that the `IndexStore` is
+    /// compatible with. For example, the `IndexStore` for the default index
+    /// returns "default".
     fn name(&self) -> &str;
 
+    /// Returns the index at the specified operation.
     fn get_index_at_op(
         &self,
         op: &Operation,
         store: &Arc<Store>,
     ) -> Result<Box<dyn ReadonlyIndex>, IndexReadError>;
 
+    /// Writes `index` to the index store and returns a read-only version of the
+    /// index.
     fn write_index(
         &self,
         index: Box<dyn MutableIndex>,
         op_id: &OperationId,
     ) -> Result<Box<dyn ReadonlyIndex>, IndexWriteError>;
 }
 
+/// Defines the interface for types that provide an index of the commits in a
+/// repository by [`CommitId`].
 pub trait Index: Send + Sync {
+    /// Returns the minimum prefix length to disambiguate `commit_id` from other
+    /// commits in the index. The length returned is the number of hexadecimal
+    /// digits in the minimum prefix.
+    ///
+    /// If the given `commit_id` doesn't exist, zero is returned since a prefix
+    /// length of zero never matches any commit id.
```

Comment:
A prefix of length zero matches every commit, right?

The original in `default_index/composite.rs` says this:

```
/// If the given `commit_id` doesn't exist, this will return the prefix
/// length that never matches with any commit ids.
```

That seems to suggest that if the index has commits "aaaa" and "aabb" and we pass in "aaab" to this function, it will return 4, since we need all 4 digits to distinguish it from "aaaa". If we instead passed in "bbbb", it would return 1, since that's enough to make it unique.

Reply:
I think I got a little confused by the implementation for `CompositeIndex`.

https://github.com/martinvonz/jj/blob/415ae5c0f9ed4d45df564a4aca6d327485a445bf/lib/src/default_index/composite.rs#L413-L425

The comment says it will return a length that never matches with any commit IDs. The "never" suggested to me that it would return a special value that never matches any commit, regardless of what commit IDs exist currently. That, combined with the `unwrap_or(0)`, made me think that zero is a special value.

And I think zero will actually be returned for composite index since `resolve_neighbor_commit_ids` will return `(None, None)` if the commit doesn't exist, right?

Reply:
"a length that never matches" is the minimum prefix length within the set `stored_commit_ids | provided_commit_id`.

> zero will actually be returned

If the index has no content, 0 will be returned. It's the minimal prefix length to disambiguate the provided `commit_id`.

Reply:
Ok, I guess I didn't fully understand how the lookup works. I will rewrite this part to say:

```
/// If the given `commit_id` doesn't exist, returns the minimum prefix
/// length which matches none of the commits in the index.
```

How's that?

Reply:
SGTM

Reply:
Ok thanks for bearing with me!